readthedocs/doc_builder/director.py

```diff
@@ -194,17 +203,36 @@ def build(self):
 
     # VCS checkout
     def checkout(self):
-        log.info(
-            "Clonning repository.",
-        )
+        """Checkout Git repo and load build config file."""
+
+        log.info("Cloning repository.")
         self.vcs_repository.update()
 
         identifier = self.data.build_commit or self.data.version.identifier
         log.info("Checking out.", identifier=identifier)
         self.vcs_repository.checkout(identifier)
 
-        self.data.config = load_yaml_config(version=self.data.version)
+        # The director is responsible for understanding which config file to use for a build.
+        # Reproducibility: Build already has a config_file set.
+        non_default_config_file = self.data.build.get("build_config_file", None)
+
+        # This logic can be extended with version-specific config files
+        if not non_default_config_file and self.data.version.project.build_config_file:
+            non_default_config_file = self.data.version.project.build_config_file
+
+        if non_default_config_file:
+            log.info(f"Using a non-default config file {non_default_config_file}.")
+        self.data.config = load_yaml_config(
+            version=self.data.version,
+            config_file=non_default_config_file,
+        )
         self.data.build["config"] = self.data.config.as_dict()
+        self.data.build["build_config_file"] = non_default_config_file
+
+        # Config file was successfully loaded and a custom path was used
+        if non_default_config_file and not self.data.version.build_config_file:
+            self.data.version.build_config_file = non_default_config_file
+            # TODO: Do we call self.data.version.save() here?
```

Comment:
Here is a thing I'd like to have input on.

The goal is to simply store the project-wide `build_config_file` that's applied. We aren't reading it or using it. That can come later if we want to do version-level customization.

Notice also that if we would call `version.save()`, it'd update the version and overwrite anything that's already there, so that matches the goal to just store a project-level configuration for going version-level later.

Reply:
I can also add here that there is of course an alternative path to just not store it for now and move on!

Reply:
Marking this resolved. After pondering about reproducibility and future scenarios, it seems fine to me that we're not saving anything on a Version. We are still saving it Builds, so we can achieve the same state as just unconditionally writing it to Version.

The ultimate choice for this configuration will still land on a project maintainer: If they change their project-wide setting, what consequence should it have for old versions? Are they migrating those, too? Or not? We can't do anything automatic here and it's a v2 thing anyways.